### PR TITLE
Inference with cropharvest models

### PR DESCRIPTION
**Pull request by ivanzvonkov**
This PR:
- Introduces inference class for doing inference with models trained on cropharvest type data
- Updates demo notebook to use inference class at the end  [link](https://github.com/nasaharvest/cropharvest/blob/828df32d3f7c0d5fc5515a979d8a1330b9927ef9/demo.ipynb)
- Adds full integration test which includes training, evaluation, inference

### CODE REVIEW

File: cropharvest/inference.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        device=None,
+        batch_size: int = 64,
+        normalizing_dict: Optional[Dict[str, np.ndarray]] = None,
+    ):
+        self.model = model
+        self.device = device
+        self.normalizing_dict = normalizing_dict
+        self.batch_size: int = batch_size
+
+        if hasattr(self.model, "predict_proba"):
```
Comment (gabrieltseng) [inline cropharvest/inference.py:33]: This might be more robust if its
```python
from sklearn.base import BaseEstimator

if isinstance(self.model, BaseEstimator)... 
```

File: cropharvest/inference.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    Class for running inference on a single tif file using an sklearn or
+    pytorch model (including jit).
+    """
+
+    def __init__(
+        self,
+        model,
+        device=None,
+        batch_size: int = 64,
+        normalizing_dict: Optional[Dict[str, np.ndarray]] = None,
```
Comment (gabrieltseng) [inline cropharvest/inference.py:26]: Maybe the normalizing dict can not have a default argument?

i.e. `normalizing_dict: Optional[Dict[str, np.ndarray]]` but without the `=  None`. 

This way it can be `None` but that has to be explicitly passed (instead of being an argument which can be ignored)

File: test/cropharvest/test_full_integration.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from cropharvest.datasets import CropHarvest
+from cropharvest.inference import Inference
+from pathlib import Path
+from sklearn.ensemble import RandomForestClassifier
+
+DATA_DIR = "data"
+TIF_FILE = Path(__file__).parent / "98-togo_2019-02-06_2020-02-01.tif"
+
+
+def test_full_integration():
```
Comment (gabrieltseng) [inline test/cropharvest/test_full_integration.py:10]: I appreciate the tests you've added 😍

File: cropharvest/inference.py (hunk 1 above)
Comment (ivanzvonkov) [reply cropharvest/inference.py:33]: Yes, good call

Comment (ivanzvonkov) [reply cropharvest/inference.py:33]: Actually on second thought BaseEstimators aren't guaranteed to have `predict_proba` so the original way may be the most safe.
See: https://stackoverflow.com/questions/30056331/how-to-list-all-scikit-learn-classifiers-that-support-predict-proba

File: cropharvest/inference.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            self.model_type = "sklearn"
+        elif TORCH_INSTALLED:
+            self.model_type = "pytorch"
+        else:
+            raise ModuleNotFoundError(
+                "Using PyTorch model but PyTorch is not installed. Please pip install torch"
+            )
+
+    @staticmethod
+    def start_date_from_str(path: Union[Path, str]) -> datetime:
```
[the remark this answers is not in the document]
Comment (ivanzvonkov) [reply cropharvest/inference.py:43]: Hmm I think it's actually pretty different. In `Engineer.process_test_file_with_region` only the end year is extracted from the filename and then the start date is created with a whole bunch of default values. Here the entire start date is taken from the file name

File: cropharvest/inference.py (hunk 2 above)
Comment (ivanzvonkov) [reply cropharvest/inference.py:26]: Good call

File: test/cropharvest/test_full_integration.py (hunk 3 above)
Comment (ivanzvonkov) [reply test/cropharvest/test_full_integration.py:10]: I have to do it because I break code too easily 🙈

File: cropharvest/inference.py (hunk 1 above)
Comment (gabrieltseng) [reply cropharvest/inference.py:33]: okay! this makes sense to me

**Review by gabrieltseng — APPROVED** [review]
